packages/catlog/src/tt/modelgen.rs

```diff
@@ -1,28 +1,38 @@
 //! Generate catlog models from doublett types.
 
+use std::fmt;
+
 use crate::dbl::model::DiscreteDblModel;
 use crate::dbl::model::MutDblModel;
+use crate::dbl::model::{DblModel, FgDblModel};
+use crate::one::category::FgCategory;
+use crate::tt::stx::MorphismType;
 use crate::tt::toplevel::Toplevel;
 use crate::tt::{eval::*, prelude::*, val::*};
+use crate::zero::QualifiedLabel;
 use crate::zero::QualifiedName;
 
 /// Generate a discrete double model from a type.
 ///
 /// Precondition: `ty` must be valid in the empty context.
-pub fn generate(toplevel: &Toplevel, ty: &TyV) -> DiscreteDblModel {
+pub fn generate(
+    toplevel: &Toplevel,
+    ty: &TyV,
+) -> (DiscreteDblModel, HashMap<QualifiedName, QualifiedLabel>) {
     let eval = Evaluator::new(toplevel, Env::Nil, 0);
-    let (elt, eval) = eval.bind_neu(text_seg("self"), ty.clone());
+    let (elt, eval) = eval.bind_self(ty.clone());
     let elt = eval.eta_neu(&elt, ty);
     let mut out = DiscreteDblModel::new(toplevel.theory.clone());
-    extract_to(&eval, &mut out, vec![], &elt, ty);
-    out
+    let mut name_translation = HashMap::new();
```

Comment:
Not sure how all the pieces will ultimately fit together, but noting in case you didn't see that there is a [`Namespace`](https://github.com/ToposInstitute/CatColab/blob/b889f9daa5ce68cca05de6f87d149f9edee064d0/packages/catlog/src/zero/qualified.rs#L352) type in `crate::zero::qualified` to do bidirectional lookups between qualified names and qualified labels. Currently, as a result of incremental refactoring, it is used only at the `catlog-wasm` level, but we might want to put that data somewhere else.

Reply:
Ah, excellent, yes I will use that Namespace type!

Reply:
Looking more into this, I don't really understand what's going on with `Namespace`. Specifically, why is there both a hashmap and an indexedhashcolumn? Is that for bidirectionality? Why do you only have bidirectionality part of the time? Maybe this should be an explicit enum?

Reply:
The short is answer is that it's for bidirectionality, which we need in the frontend:

- forward direction (name -> label) to *display* a name
- inverse direction (label -> name) to take a name as *input* from the user

Reply:
I grant that the data structure itself is a little weird. There might be a better way to do it.